File: Data Structures & Algorithms/coin-change/submission-2.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amt) {
        int n = coins.size(), m = amt;
        vector<vector<int>> dp (n+1, vector<int>(m+1, m+1));

        for(int i = 0;i<n+1;i++) {
            dp[i][0] = 0;
        }

        // for(int j = 0;j<m+1;j++) {
        //     if(j%coins[0] == 0)
        //         dp[1][j] = j/coins[0];
        // }

        for(int i = 1;i<n+1;i++) {
            for(int j = 1; j<m+1;j++) {
                if(coins[i-1] <= j)
                    dp[i][j] = min(1+dp[i][j-coins[i-1]], dp[i-1][j]);
                else
                    dp[i][j] = dp[i-1][j];
            }
        }

        return dp[n][amt] > amt ? -1: dp[n][amt] ;
    }
};

File: Data Structures & Algorithms/min-cost-to-connect-points/submission-2.java
class Solution {
    public int minCostConnectPoints(int[][] points) {
        int cost = 0, curr = 0, n = points.length;
        boolean[] vis = new boolean[n];
        int[] dis = new int[n];
        Arrays.fill(dis, (int)1e7);
        int cnt = 0;

        while(cnt < n - 1) {
            vis[curr] = true;
            int next = -1;
            for(int i = 0;i<n;i++) {
                if(vis[i]) continue;

                int disC = Math.abs(points[i][0]-points[curr][0]) + Math.abs(points[i][1] - points[curr][1]);
                dis[i] = Math.min(dis[i], disC);

                if(next == -1 || dis[i] < dis[next]) {
                    next = i;
                }
            }
            cost+=dis[next];
            curr = next;
            cnt++;
        }
        return cost;
    }
}
